docs: provision spark dependencies (#409)
stackabletech/spark-k8s-operator: docs/modules/spark-k8s/pages/usage-guide/job-dependencies.adoc

== Overview

> IMPORTANT: With the platform release 23.4.1 and Apache Spark 3.3.x (and all previous releases), dynamic provisioning of dependencies using the Spark `packages` field doesn't work. This is a known problem with Spark and is tracked https://github.com/stackabletech/spark-k8s-operator/issues/141[here].

Reviewer: ```suggestion
IMPORTANT: With the platform release 23.4.1 and Apache Spark 3.3.x (and all previous releases), dynamic provisioning of dependencies using the Spark `packages` field doesn't work.
This is a known problem with Spark and is tracked https://github.com/stackabletech/spark-k8s-operator/issues/141[here].
```

The container images provided by Stackable include Apache Spark and PySpark applications and libraries. In addtion they include commonly used libraries to connect to storage systems supporting the `hdfs://`, `s3a://` and `abfs://` protocols. These systems are commonly used to store data processed by Spark applications.

> Sometimes the applications need to integrate with additional systems or use processing algorithms not included in the Apache Spark distribution. This guide explains how users can provision their Spark jobs with additional dependencies to support these requirements.

Reviewer: ```suggestion
Sometimes the applications need to integrate with additional systems or use processing algorithms not included in the Apache Spark distribution.
This guide explains how you can provision your Spark jobs with additional dependencies to support these requirements.
```

Please [adress the reader as 'you'](https://developers.google.com/style/person)

There are more instances of this in the text, please fix them all. Thanks!

== Dependency provisioning

There are multiple ways to submit Apache Spark jobs with external dependencies. Each has its own advantages and disadvantages and the choice of one over the other depends on existing technical and managerial constraints.

To provision job dependencies in their workloads, users have to construct their `SparkApplication` with one of the following dependency specifications:

> - Custom Spark images

Reviewer: Use `*` instead of `-` for bullet point lists.

|Guaranteed
|Small to Medium

|Maven/Java packages
|Small
|Not guaranteed
|Small

|Python packages
|Small
|Not guaranteed
|Small
|===

=== Custom Spark images

> With this method, users submit a `SparkApplication` for which the `sparkImage` refers to the full custom image name. It is recommended the users base their image on one of the Stackable images to ensure compatibility with the Stackable operator.

Reviewer: ```suggestion
With this method, you submit a SparkApplication for which the `sparkImage` refers to the full custom image name.
It is recommended you base your image on one of the Stackable images to ensure compatibility with the Stackable operator.
```

Don't use code style with 'kinds' like SparkApplication or StatefulSet

    spark.sql.catalog.local.warehouse: /tmp/warehouse
  deps:
    packages:
      - org.apache.iceberg:iceberg-spark-runtime-3.4_2.12:1.3.1 # <1>
...
----

<1> Maven package coordinates for Apache Iceberg. This is downloaded from the Manven repository and made available to the Spark application.

> IMPORTANT: Currently it's not possible to provision dependencies that are loaded by the JVM's (system class loader)[https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ClassLoader.html#getSystemClassLoader()]. Such dependencies include JDBC drivers. If you need access to JDBC sources from your Spark application, consider building your own custom Spark image as shown above.

Reviewer: ```suggestion
IMPORTANT: Currently it's not possible to provision dependencies that are loaded by the JVM's https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ClassLoader.html#getSystemClassLoader()[system class loader].
Such dependencies include JDBC drivers.
If you need access to JDBC sources from your Spark application, consider building your own custom Spark image as shown above.
```

incorrect link syntax used.
